File: test/keyspace_test.py
#!/usr/bin/env python
#
# Copyright 2017 Google Inc.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
import json

from vtdb import vtgate_client

import environment
import tablet
import utils

SHARDED_KEYSPACE = 'test_keyspace_sharded'
UNSHARDED_KEYSPACE = 'test_keyspace_unsharded'

# shards for SHARDED_KEYSPACE
# range '' - 80
shard_0_master = tablet.Tablet()
shard_0_replica = tablet.Tablet()
# range 80 - ''
shard_1_master = tablet.Tablet()
shard_1_replica = tablet.Tablet()

# shard for UNSHARDED_KEYSPACE
unsharded_master = tablet.Tablet()
unsharded_replica = tablet.Tablet()

shard_names = ['-80', '80-']
shard_kid_map = {
    '-80': [527875958493693904, 626750931627689502,
            345387386794260318, 332484755310826578,
            1842642426274125671, 1326307661227634652,
            1761124146422844620, 1661669973250483744,
            3361397649937244239, 2444880764308344533],
    '80-': [9767889778372766922, 9742070682920810358,
            10296850775085416642, 9537430901666854108,
            10440455099304929791, 11454183276974683945,
            11185910247776122031, 10460396697869122981,
            13379616110062597001, 12826553979133932576],
}

create_vt_insert_test = '''create table vt_insert_test (
id bigint auto_increment,
msg varchar(64),
keyspace_id bigint(20) unsigned NOT NULL,
primary key (id)
) Engine=InnoDB'''


def setUpModule():
  try:
    environment.topo_server().setup()

    setup_procs = [
        shard_0_master.init_mysql(),
        shard_0_replica.init_mysql(),
        shard_1_master.init_mysql(),
        shard_1_replica.init_mysql(),
        unsharded_master.init_mysql(),
        unsharded_replica.init_mysql(),
        ]
    utils.wait_procs(setup_procs)
    setup_tablets()
  except:
    tearDownModule()
    raise


def tearDownModule():
  utils.required_teardown()
  if utils.options.skip_teardown:
    return

  tablet.kill_tablets([shard_0_master, shard_0_replica,
                       shard_1_master, shard_1_replica])
  teardown_procs = [
      shard_0_master.teardown_mysql(),
      shard_0_replica.teardown_mysql(),
      shard_1_master.teardown_mysql(),
      shard_1_replica.teardown_mysql(),
      unsharded_master.teardown_mysql(),
      unsharded_replica.teardown_mysql(),
      ]
  utils.wait_procs(teardown_procs, raise_on_error=False)

  environment.topo_server().teardown()
  utils.kill_sub_processes()
  utils.remove_tmp_files()

  shard_0_master.remove_tree()
  shard_0_replica.remove_tree()
  shard_1_master.remove_tree()
  shard_1_replica.remove_tree()
  unsharded_master.remove_tree()
  unsharded_replica.remove_tree()


def setup_tablets():
  setup_sharded_keyspace()
  setup_unsharded_keyspace()
  utils.VtGate().start(tablets=[
      shard_0_master, shard_0_replica,
      shard_1_master, shard_1_replica,
      unsharded_master, unsharded_replica,
      ])
  utils.vtgate.wait_for_endpoints(
      '%s.%s.master' % (SHARDED_KEYSPACE, '80-'),
      1)
  utils.vtgate.wait_for_endpoints(
      '%s.%s.replica' % (SHARDED_KEYSPACE, '80-'),
      1)
  utils.vtgate.wait_for_endpoints(
      '%s.%s.master' % (SHARDED_KEYSPACE, '-80'),
      1)
  utils.vtgate.wait_for_endpoints(
      '%s.%s.replica' % (SHARDED_KEYSPACE, '-80'),
      1)
  utils.vtgate.wait_for_endpoints(
      '%s.%s.master' % (UNSHARDED_KEYSPACE, '0'),
      1)
  utils.vtgate.wait_for_endpoints(
      '%s.%s.replica' % (UNSHARDED_KEYSPACE, '0'),
      1)


def setup_sharded_keyspace():
  utils.run_vtctl(['CreateKeyspace', SHARDED_KEYSPACE])
  utils.run_vtctl(['SetKeyspaceShardingInfo', '-force', SHARDED_KEYSPACE,
                   'keyspace_id', 'uint64'])

  shard_0_master.init_tablet(
      'replica',
      keyspace=SHARDED_KEYSPACE,
      shard='-80',
      tablet_index=0)
  shard_0_replica.init_tablet(
      'replica',
      keyspace=SHARDED_KEYSPACE,
      shard='-80',
      tablet_index=1)
  shard_1_master.init_tablet(
      'replica',
      keyspace=SHARDED_KEYSPACE,
      shard='80-',
      tablet_index=0)
  shard_1_replica.init_tablet(
      'replica',
      keyspace=SHARDED_KEYSPACE,
      shard='80-',
      tablet_index=1)

  for t in [shard_0_master, shard_0_replica, shard_1_master, shard_1_replica]:
    t.create_db('vt_test_keyspace_sharded')
    t.mquery(shard_0_master.dbname, create_vt_insert_test)
    t.start_vttablet(wait_for_state=None)

  for t in [shard_0_master, shard_0_replica, shard_1_master, shard_1_replica]:
    t.wait_for_vttablet_state('NOT_SERVING')

  utils.run_vtctl(['InitShardMaster', '-force', '%s/-80' % SHARDED_KEYSPACE,
                   shard_0_master.tablet_alias], auto_log=True)
  utils.run_vtctl(['InitShardMaster', '-force', '%s/80-' % SHARDED_KEYSPACE,
                   shard_1_master.tablet_alias], auto_log=True)

  for t in [shard_0_replica, shard_1_replica]:
    utils.wait_for_tablet_type(t.tablet_alias, 'replica')
  for t in [shard_0_master, shard_0_replica, shard_1_master, shard_1_replica]:
    t.wait_for_vttablet_state('SERVING')

  # rebuild to be sure we have the latest data
  utils.run_vtctl(
      ['RebuildKeyspaceGraph', SHARDED_KEYSPACE], auto_log=True)
  utils.check_srv_keyspace('test_nj', SHARDED_KEYSPACE,
                           'Partitions(master): -80 80-\n'
                           'Partitions(rdonly): -80 80-\n'
                           'Partitions(replica): -80 80-\n')


def setup_unsharded_keyspace():
  utils.run_vtctl(['CreateKeyspace', UNSHARDED_KEYSPACE])
  utils.run_vtctl(['SetKeyspaceShardingInfo', '-force', UNSHARDED_KEYSPACE,
                   'keyspace_id', 'uint64'])

  unsharded_master.init_tablet(
      'replica',
      keyspace=UNSHARDED_KEYSPACE,
      shard='0',
      tablet_index=0)
  unsharded_replica.init_tablet(
      'replica',
      keyspace=UNSHARDED_KEYSPACE,
      shard='0',
      tablet_index=1)

  for t in [unsharded_master, unsharded_replica]:
    t.create_db('vt_test_keyspace_unsharded')
    t.mquery(unsharded_master.dbname, create_vt_insert_test)
    t.start_vttablet(wait_for_state=None)

  for t in [unsharded_master, unsharded_replica]:
    t.wait_for_vttablet_state('NOT_SERVING')

  utils.run_vtctl(['InitShardMaster', '-force', '%s/0' % UNSHARDED_KEYSPACE,
                   unsharded_master.tablet_alias], auto_log=True)

  for t in [unsharded_replica]:
    utils.wait_for_tablet_type(t.tablet_alias, 'replica')
  for t in [unsharded_master, unsharded_replica]:
    t.wait_for_vttablet_state('SERVING')

  # rebuild to be sure we have the right version
  utils.run_vtctl(['RebuildKeyspaceGraph', UNSHARDED_KEYSPACE], auto_log=True)
  utils.check_srv_keyspace('test_nj', UNSHARDED_KEYSPACE,
                           'Partitions(master): -\n'
                           'Partitions(rdonly): -\n'
                           'Partitions(replica): -\n')


ALL_DB_TYPES = ['master', 'rdonly', 'replica']


class TestKeyspace(unittest.TestCase):

  def _read_srv_keyspace(self, keyspace_name):
    protocol, addr = utils.vtgate.rpc_endpoint(python=True)
    conn = vtgate_client.connect(protocol, addr, 30.0)
    result = conn.get_srv_keyspace(keyspace_name)
    conn.close()
    return result

  def test_get_keyspace(self):
    ki = utils.run_vtctl_json(['GetKeyspace', UNSHARDED_KEYSPACE])
    self.assertEqual('keyspace_id', ki['sharding_column_name'])
    self.assertEqual(1, ki['sharding_column_type'])

  def test_delete_keyspace(self):
    utils.run_vtctl(['CreateKeyspace', 'test_delete_keyspace'])
    utils.run_vtctl(['CreateShard', 'test_delete_keyspace/0'])
    utils.run_vtctl(
        ['InitTablet', '-keyspace=test_delete_keyspace', '-shard=0',
         'test_nj-0000000100', 'master'])

    # Can't delete keyspace if there are shards present.
    utils.run_vtctl(
        ['DeleteKeyspace', 'test_delete_keyspace'], expect_fail=True)
    # Can't delete shard if there are tablets present.
    utils.run_vtctl(['DeleteShard', '-even_if_serving',
                     'test_delete_keyspace/0'], expect_fail=True)

    # Use recursive DeleteShard to remove tablets.
    utils.run_vtctl(['DeleteShard', '-even_if_serving', '-recursive',
                     'test_delete_keyspace/0'])
    # Now non-recursive DeleteKeyspace should work.
    utils.run_vtctl(['DeleteKeyspace', 'test_delete_keyspace'])

    # Start over and this time use recursive DeleteKeyspace to do everything.
    utils.run_vtctl(['CreateKeyspace', 'test_delete_keyspace'])
    utils.run_vtctl(['CreateShard', 'test_delete_keyspace/0'])
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=0', 'test_nj-0000000100', 'master'])

    # Create the serving/replication entries and check that they exist,
    # so we can later check they're deleted.
    utils.run_vtctl(['RebuildKeyspaceGraph', 'test_delete_keyspace'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'])
    utils.run_vtctl(['GetSrvKeyspace', 'test_nj', 'test_delete_keyspace'])

    # Recursive DeleteKeyspace
    utils.run_vtctl(['DeleteKeyspace', '-recursive', 'test_delete_keyspace'])

    # Check that everything is gone.
    utils.run_vtctl(['GetKeyspace', 'test_delete_keyspace'], expect_fail=True)
    utils.run_vtctl(['GetShard', 'test_delete_keyspace/0'], expect_fail=True)
    utils.run_vtctl(['GetTablet', 'test_nj-0000000100'], expect_fail=True)
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'],
        expect_fail=True)
    utils.run_vtctl(
        ['GetSrvKeyspace', 'test_nj', 'test_delete_keyspace'],
        expect_fail=True)

  def test_remove_keyspace_cell(self):
    utils.run_vtctl(['CreateKeyspace', 'test_delete_keyspace'])
    utils.run_vtctl(['CreateShard', 'test_delete_keyspace/0'])
    utils.run_vtctl(['CreateShard', 'test_delete_keyspace/1'])
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=0', 'test_ca-0000000100', 'master'])
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=1', 'test_ca-0000000101', 'master'])
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=0', 'test_nj-0000000100', 'replica'])
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=1', 'test_nj-0000000101', 'replica'])

    # Create the serving/replication entries and check that they exist,
    # so we can later check they're deleted.
    utils.run_vtctl(['RebuildKeyspaceGraph', 'test_delete_keyspace'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/1'])
    utils.run_vtctl(['GetSrvKeyspace', 'test_nj', 'test_delete_keyspace'])
    utils.run_vtctl(['GetSrvKeyspace', 'test_ca', 'test_delete_keyspace'])

    # Just remove the shard from one cell (including tablets),
    # but leaving the global records and other cells/shards alone.
    utils.run_vtctl(
        ['RemoveShardCell', '-recursive', 'test_delete_keyspace/0', 'test_nj'])
    # Check that the shard is gone from test_nj.
    srv_keyspace = utils.run_vtctl_json(['GetSrvKeyspace', 'test_nj', 'test_delete_keyspace'])
    for partition in srv_keyspace['partitions']:
      self.assertEqual(len(partition['shard_references']), 1,
          'RemoveShardCell should have removed one shard from the target cell: ' +
          json.dumps(srv_keyspace))
    utils.run_vtctl(['RebuildKeyspaceGraph', 'test_delete_keyspace'])

    utils.run_vtctl(['GetKeyspace', 'test_delete_keyspace'])
    utils.run_vtctl(['GetShard', 'test_delete_keyspace/0'])
    utils.run_vtctl(['GetTablet', 'test_ca-0000000100'])
    utils.run_vtctl(['GetTablet', 'test_nj-0000000100'], expect_fail=True)
    utils.run_vtctl(['GetTablet', 'test_nj-0000000101'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_ca', 'test_delete_keyspace/0'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'],
        expect_fail=True)
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/1'])
    utils.run_vtctl(['GetSrvKeyspace', 'test_nj', 'test_delete_keyspace'])

    # Add it back to do another test.
    utils.run_vtctl(
        ['InitTablet', '-port=1234', '-keyspace=test_delete_keyspace',
         '-shard=0', 'test_nj-0000000100', 'replica'])
    utils.run_vtctl(['RebuildKeyspaceGraph', 'test_delete_keyspace'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'])

    # Now use RemoveKeyspaceCell to remove all shards.
    utils.run_vtctl(
        ['RemoveKeyspaceCell', '-recursive', 'test_delete_keyspace',
         'test_nj'])
    utils.run_vtctl(['RebuildKeyspaceGraph', 'test_delete_keyspace'])

    utils.run_vtctl(
        ['GetShardReplication', 'test_ca', 'test_delete_keyspace/0'])
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/0'],
        expect_fail=True)
    utils.run_vtctl(
        ['GetShardReplication', 'test_nj', 'test_delete_keyspace/1'],
        expect_fail=True)

    # Clean up.
    utils.run_vtctl(['DeleteKeyspace', '-recursive', 'test_delete_keyspace'])

  def test_shard_count(self):
    sharded_ks = self._read_srv_keyspace(SHARDED_KEYSPACE)
    for db_type in ALL_DB_TYPES:
      self.assertEqual(sharded_ks.get_shard_count(db_type), 2)
    unsharded_ks = self._read_srv_keyspace(UNSHARDED_KEYSPACE)
    for db_type in ALL_DB_TYPES:
      self.assertEqual(unsharded_ks.get_shard_count(db_type), 1)

  def test_shard_names(self):
    sharded_ks = self._read_srv_keyspace(SHARDED_KEYSPACE)
    for db_type in ALL_DB_TYPES:
      self.assertEqual(sharded_ks.get_shard_names(db_type), ['-80', '80-'])
    unsharded_ks = self._read_srv_keyspace(UNSHARDED_KEYSPACE)
    for db_type in ALL_DB_TYPES:
      self.assertEqual(unsharded_ks.get_shard_names(db_type), ['0'])

  def test_keyspace_id_to_shard_name(self):
    # test all keyspace_id in a sharded keyspace go to the right shard
    sharded_ks = self._read_srv_keyspace(SHARDED_KEYSPACE)
    for sn in shard_names:
      for keyspace_id in shard_kid_map[sn]:
        self.assertEqual(
            sharded_ks.keyspace_id_to_shard_name_for_db_type(keyspace_id,
                                                             'master'), sn)

    # take all keyspace_ids, make sure for unsharded they stay on'0'
    unsharded_ks = self._read_srv_keyspace(UNSHARDED_KEYSPACE)
    for sn in shard_names:
      for keyspace_id in shard_kid_map[sn]:
        self.assertEqual(
            unsharded_ks.keyspace_id_to_shard_name_for_db_type(
                keyspace_id, 'master'),
            '0')

  def test_get_srv_keyspace_names(self):
    stdout, _ = utils.run_vtctl(['GetSrvKeyspaceNames', 'test_nj'],
                                trap_output=True)
    self.assertEqual(
        set(stdout.splitlines()), {SHARDED_KEYSPACE, UNSHARDED_KEYSPACE})


if __name__ == '__main__':
  utils.main()

File: go/test/endtoend/keyspace/keyspace_test.go
/*
Copyright 2019 The Vitess Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package sequence

import (
	"context"
	"encoding/binary"
	"path"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"vitess.io/vitess/go/constants/sidecar"
	"vitess.io/vitess/go/json2"
	"vitess.io/vitess/go/vitesst"
	"vitess.io/vitess/go/vt/key"
	"vitess.io/vitess/go/vt/topo"
	_ "vitess.io/vitess/go/vt/topo/etcd2topo"
	"vitess.io/vitess/go/vt/topo/topoproto"
	"vitess.io/vitess/go/vt/vtctl/reparentutil/policy"

	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
	vtctldatapb "vitess.io/vitess/go/vt/proto/vtctldata"
)

var (
	clusterForKSTest      *vitesst.Cluster
	keyspaceShardedName   = "test_ks_sharded"
	keyspaceUnshardedName = "test_ks_unsharded"
	cell                  = "zone1"
	cell2                 = "zone2"
	servedTypes           = map[topodatapb.TabletType]bool{topodatapb.TabletType_PRIMARY: true, topodatapb.TabletType_REPLICA: true, topodatapb.TabletType_RDONLY: true}
	sqlSchema             = `create table vt_insert_test (
								id bigint auto_increment,
								msg varchar(64),
								keyspace_id bigint(20) unsigned NOT NULL,
								primary key (id)
								) Engine=InnoDB`
	vSchema = `{
      "sharded": true,
      "vindexes": {
        "hash_index": {
          "type": "hash"
        }
      },
      "tables": {
        "vt_insert_test": {
           "column_vindexes": [
            {
              "column": "keyspace_id",
              "name": "hash_index"
            }
          ]
        }
      }
	}`
	shardKIdMap = map[string][]uint64{
		"-80": {
			527875958493693904, 626750931627689502,
			345387386794260318, 332484755310826578,
			1842642426274125671, 1326307661227634652,
			1761124146422844620, 1661669973250483744,
			3361397649937244239, 2444880764308344533,
		},
		"80-": {
			9767889778372766922, 9742070682920810358,
			10296850775085416642, 9537430901666854108,
			10440455099304929791, 11454183276974683945,
			11185910247776122031, 10460396697869122981,
			13379616110062597001, 12826553979133932576,
		},
	}
)

func setup(t *testing.T) {
	t.Helper()
	ctx := t.Context()

	cluster, err := vitesst.NewCluster(
		vitesst.WithCells(cell, cell2),
		vitesst.WithKeyspace(keyspaceShardedName).
			WithShardNames("-80", "80-").
			WithReplicas(1).
			WithSchema(sqlSchema).
			WithVSchema(vSchema),
		vitesst.WithKeyspace(keyspaceUnshardedName).
			WithShardNames(keyspaceUnshardedName).
			WithReplicas(1).
			WithSchema(sqlSchema),
	)
	require.NoError(t, err)

	cleanup, err := cluster.Start(ctx)
	t.Cleanup(func() {
		cleanupCtx, cancel := context.WithTimeout(context.WithoutCancel(t.Context()), time.Minute)
		defer cancel()
		if t.Failed() {
			cluster.DumpDiagnostics(cleanupCtx, t.Logf)
		}
		if err := cleanup(cleanupCtx); err != nil {
			t.Logf("cluster teardown: %v", err)
		}
	})
	require.NoError(t, err)

	clusterForKSTest = cluster
	require.NoError(t, cluster.Vtctld().ExecuteCommand(ctx, "RebuildKeyspaceGraph", keyspaceShardedName))
	require.NoError(t, cluster.Vtctld().ExecuteCommand(ctx, "RebuildKeyspaceGraph", keyspaceUnshardedName))
}

// TestDurabilityPolicyField tests that the DurabilityPolicy field of a keyspace can be set during creation, read and updated later
// from vtctld server and the vtctl binary
func TestDurabilityPolicyField(t *testing.T) {
	setup(t)
	ctx := t.Context()

	out, err := clusterForKSTest.Vtctld().ExecuteCommandWithOutput(ctx, "CreateKeyspace", "ks_durability", "--durability-policy=semi_sync")
	require.NoError(t, err, out)
	checkDurabilityPolicy(t, policy.DurabilitySemiSync)

	out, err = clusterForKSTest.Vtctld().ExecuteCommandWithOutput(ctx, "SetKeyspaceDurabilityPolicy", "ks_durability", "--durability-policy=none")
	require.NoError(t, err, out)
	checkDurabilityPolicy(t, policy.DurabilityNone)

	out, err = clusterForKSTest.Vtctld().ExecuteCommandWithOutput(ctx, "DeleteKeyspace", "ks_durability")
	require.NoError(t, err, out)

	out, err = clusterForKSTest.Vtctld().ExecuteCommandWithOutput(ctx, "CreateKeyspace", "--durability-policy=semi_sync", "ks_durability")
	require.NoError(t, err, out)
	checkDurabilityPolicy(t, policy.DurabilitySemiSync)

	out, err = clusterForKSTest.Vtctld().ExecuteCommandWithOutput(ctx, "DeleteKeyspace", "ks_durability")
	require.NoError(t, err, out)
}

func checkDurabilityPolicy(t *testing.T, durabilityPolicy string) {
	ks, err := getKeyspace(t, "ks_durability")
	require.NoError(t, err)
	require.Equal(t, ks.Keyspace.DurabilityPolicy, durabilityPolicy)
}

func TestGetSrvKeyspaceNames(t *testing.T) {
	setup(t)
	data, err := clusterForKSTest.Vtctld().ExecuteCommandWithOutput(t.Context(), "GetSrvKeyspaceNames", cell)
	require.Nil(t, err)

	namesByCell := map[string]*vtctldatapb.GetSrvKeyspaceNamesResponse_NameList{}
	err = json2.Unmarshal([]byte(data), &namesByCell)
	require.NoError(t, err)

	assert.Contains(t, namesByCell[cell].Names, keyspaceUnshardedName)
	assert.Contains(t, namesByCell[cell].Names, keyspaceShardedName)
}

func TestGetSrvKeyspacePartitions(t *testing.T) {
	setup(t)
	shardedSrvKeyspace := getSrvKeyspace(t, cell, keyspaceShardedName)
	otherShardRefFound := false
	for _, partition := range shardedSrvKeyspace.Partitions {
		if servedTypes[partition.ServedType] {
			for _, shardRef := range partition.ShardReferences {
				assert.True(t, shardRef.Name == "-80" || shardRef.Name == "80-")
			}
		} else {
			otherShardRefFound = true
		}
	}
	assert.True(t, !otherShardRefFound)

	unShardedSrvKeyspace := getSrvKeyspace(t, cell, keyspaceUnshardedName)
	otherShardRefFound = false
	for _, partition := range unShardedSrvKeyspace.Partitions {
		if servedTypes[partition.ServedType] {
			for _, shardRef := range partition.ShardReferences {
				assert.True(t, shardRef.Name == keyspaceUnshardedName)
			}
		} else {
			otherShardRefFound = true
		}
	}
	assert.True(t, !otherShardRefFound)
}

func TestShardNames(t *testing.T) {
	setup(t)
	output, err := getSrvKeyspaces(t, keyspaceShardedName, cell)
	require.NoError(t, err)
	require.NotNil(t, output[cell], "no srvkeyspace for cell %s", cell)
}

func TestGetKeyspace(t *testing.T) {
	setup(t)
	_, err := getKeyspace(t, keyspaceUnshardedName)
	require.Nil(t, err)
}

func TestDeleteKeyspace(t *testing.T) {
	setup(t)
	ctx := t.Context()
	vtctld := clusterForKSTest.Vtctld()

	_ = vtctld.ExecuteCommand(ctx, "CreateKeyspace", "test_delete_keyspace", "--sidecar-db-name", sidecar.DefaultName)
	_ = vtctld.ExecuteCommand(ctx, "CreateShard", "test_delete_keyspace/0")
	_ = initTablet(ctx, cell, 999, topodatapb.TabletType_PRIMARY, 0, "test_delete_keyspace", "0")

	// Can't delete keyspace if there are shards present.
	err := vtctld.ExecuteCommand(ctx, "DeleteKeyspace", "test_delete_keyspace")
	require.Error(t, err)

	// Can't delete shard if there are tablets present.
	err = vtctld.ExecuteCommand(ctx, "DeleteShards", "--even-if-serving", "test_delete_keyspace/0")
	require.Error(t, err)

	// Use recursive DeleteShard to remove tablets.
	_ = vtctld.ExecuteCommand(ctx, "DeleteShards", "--even-if-serving", "--recursive", "test_delete_keyspace/0")
	// Now non-recursive DeleteKeyspace should work.
	_ = vtctld.ExecuteCommand(ctx, "DeleteKeyspace", "test_delete_keyspace")

	// Start over and this time use recursive DeleteKeyspace to do everything.
	_ = vtctld.ExecuteCommand(ctx, "CreateKeyspace", "test_delete_keyspace", "--sidecar-db-name", sidecar.DefaultName)
	_ = vtctld.ExecuteCommand(ctx, "CreateShard", "test_delete_keyspace/0")
	_ = initTablet(ctx, cell, 999, topodatapb.TabletType_PRIMARY, 1234, "test_delete_keyspace", "0")

	// Create the serving/replication entries and check that they exist,
	//  so we can later check they're deleted.
	_ = vtctld.ExecuteCommand(ctx, "RebuildKeyspaceGraph", "test_delete_keyspace")
	_, _ = vtctld.ExecuteCommandWithOutput(ctx, "GetShardReplication", "test_delete_keyspace/0", cell)
	_ = vtctld.ExecuteCommand(ctx, "GetSrvKeyspace", cell, "test_delete_keyspace")

	// Recursive DeleteKeyspace
	_ = vtctld.ExecuteCommand(ctx, "DeleteKeyspace", "--recursive", "test_delete_keyspace")

	// Check that everything is gone.
	err = vtctld.ExecuteCommand(ctx, "GetKeyspace", "test_delete_keyspace")
	require.Error(t, err)
	err = vtctld.ExecuteCommand(ctx, "GetShard", "test_delete_keyspace/0")
	require.Error(t, err)
	err = vtctld.ExecuteCommand(ctx, "GetTablet", "zone1-0000000999")
	require.Error(t, err)
	_, err = vtctld.ExecuteCommandWithOutput(ctx, "GetShardReplication", "test_delete_keyspace/0", cell)
	require.Error(t, err)
	err = vtctld.ExecuteCommand(ctx, "GetSrvKeyspace", cell, "test_delete_keyspace")
	require.Error(t, err)
	ksMap, err := getSrvKeyspaces(t, "test_delete_keyspace", cell)
	require.NoError(t, err)
	require.Empty(t, ksMap[cell])
}

func TestShardCountForAllKeyspaces(t *testing.T) {
	setup(t)
	testShardCountForKeyspace(t, keyspaceUnshardedName, 1)
	testShardCountForKeyspace(t, keyspaceShardedName, 2)
}

func testShardCountForKeyspace(t *testing.T, keyspace string, count int) {
	srvKeyspace := getSrvKeyspace(t, cell, keyspace)

	// for each served type PRIMARY REPLICA RDONLY, the shard ref count should match
	for _, partition := range srvKeyspace.Partitions {
		if servedTypes[partition.ServedType] {
			assert.Equal(t, len(partition.ShardReferences), count)
		}
	}
}

func TestShardNameForAllKeyspaces(t *testing.T) {
	setup(t)
	testShardNameForKeyspace(t, keyspaceUnshardedName, []string{"test_ks_unsharded"})
	testShardNameForKeyspace(t, keyspaceShardedName, []string{"-80", "80-"})
}

func testShardNameForKeyspace(t *testing.T, keyspace string, shardNames []string) {
	srvKeyspace := getSrvKeyspace(t, cell, keyspace)

	// for each served type PRIMARY REPLICA RDONLY, the shard ref count should match
	for _, partition := range srvKeyspace.Partitions {
		if servedTypes[partition.ServedType] {
			for _, shardRef := range partition.ShardReferences {
				assert.Contains(t, shardNames, shardRef.Name)
			}
		}
	}
}

func TestKeyspaceToShardName(t *testing.T) {
	setup(t)
	var id []byte
	srvKeyspace := getSrvKeyspace(t, cell, keyspaceShardedName)

	// for each served type PRIMARY REPLICA RDONLY, the shard ref count should match
	for _, partition := range srvKeyspace.Partitions {
		if partition.ServedType == topodatapb.TabletType_PRIMARY {
			for _, shardRef := range partition.ShardReferences {
				shardKIDs := shardKIdMap[shardRef.Name]
				for _, kid := range shardKIDs {
					id = packKeyspaceID(kid)
					assert.True(t, key.Compare(shardRef.KeyRange.Start, id) <= 0 &&
						(key.Empty(shardRef.KeyRange.End) || key.Compare(id, shardRef.KeyRange.End) < 0))
				}
			}
		}
	}

	srvKeyspace = getSrvKeyspace(t, cell, keyspaceUnshardedName)

	for _, partition := range srvKeyspace.Partitions {
		if partition.ServedType == topodatapb.TabletType_PRIMARY {
			for _, shardRef := range partition.ShardReferences {
				assert.Equal(t, shardRef.Name, keyspaceUnshardedName)
			}
		}
	}
}

// packKeyspaceID packs this into big-endian and returns byte[] to do a byte-wise comparison.
func packKeyspaceID(keyspaceID uint64) []byte {
	var keybytes [8]byte
	binary.BigEndian.PutUint64(keybytes[:], keyspaceID)
	return keybytes[:]
}

func getSrvKeyspace(t *testing.T, cell string, ksname string) *topodatapb.SrvKeyspace {
	output, err := getSrvKeyspaces(t, ksname, cell)
	require.NoError(t, err)

	srvKeyspace := output[cell]
	require.NotNil(t, srvKeyspace, "no srvkeyspace for cell %s", cell)

	return srvKeyspace
}

// getSrvKeyspaces returns a mapping of cell to srv keyspace for the given keyspace.
func getSrvKeyspaces(t *testing.T, keyspace string, cells ...string) (map[string]*topodatapb.SrvKeyspace, error) {
	args := append([]string{"GetSrvKeyspaces", keyspace}, cells...)
	out, err := clusterForKSTest.Vtctld().ExecuteCommandWithOutput(t.Context(), args...)
	if err != nil {
		return nil, err
	}

	ksMap := map[string]*topodatapb.SrvKeyspace{}
	err = json2.Unmarshal([]byte(out), &ksMap)
	return ksMap, err
}

// getKeyspace fetches a keyspace record and parses the response.
func getKeyspace(t *testing.T, keyspace string) (*vtctldatapb.Keyspace, error) {
	data, err := clusterForKSTest.Vtctld().ExecuteCommandWithOutput(t.Context(), "GetKeyspace", keyspace)
	if err != nil {
		return nil, err
	}

	var ks vtctldatapb.Keyspace
	if err := json2.UnmarshalPB([]byte(data), &ks); err != nil {
		return nil, err
	}
	return &ks, nil
}

// initTablet creates a bare tablet record in the topology server, without a
// running mysqld or vttablet. The record and its shard replication entry live
// in the cell's own topo root, which the Vitess components reach through the
// in-network cell address. That address is not reachable from the host, so the
// shared etcd is opened at the cell's data root directly and the records are
// written there.
func initTablet(ctx context.Context, cell string, uid uint32, tabletType topodatapb.TabletType, httpPort int32, keyspace, shard string) error {
	etcdAddr, err := clusterForKSTest.EtcdAddr(ctx)
	if err != nil {
		return err
	}
	ts, err := topo.OpenServer("etcd2", etcdAddr, "/vitess/"+cell)
	if err != nil {
		return err
	}
	defer ts.Close()

	conn, err := ts.ConnForCell(ctx, topo.GlobalCell)
	if err != nil {
		return err
	}

	alias := &topodatapb.TabletAlias{
		Cell: cell,
		Uid:  uid,
	}
	tablet := &topodatapb.Tablet{
		Alias:    alias,
		Hostname: "localhost",
		Type:     tabletType,
		PortMap: map[string]int32{
			"vt": httpPort,
		},
		Keyspace: keyspace,
		Shard:    shard,
	}

	data, err := tablet.MarshalVT()
	if err != nil {
		return err
	}
	tabletPath := path.Join(topo.TabletsPath, topoproto.TabletAliasString(alias), topo.TabletFile)
	if _, err := conn.Create(ctx, tabletPath, data); err != nil {
		return err
	}

	return ts.UpdateShardReplicationFields(ctx, topo.GlobalCell, keyspace, shard,
		func(sr *topodatapb.ShardReplication) error {
			sr.Nodes = append(sr.Nodes, &topodatapb.ShardReplication_Node{TabletAlias: alias})
			return nil
		})
}
